pytorch_lightning/utilities/hpu_device.py

```diff
@@ -0,0 +1,58 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+from torch.optim.optimizer import Optimizer
+
+import pytorch_lightning as pl
+
+
+class HPUDeviceUtils:
+    """HPU helpers."""
+
+    @staticmethod
+    # Gaudi HW performs convolution operations with filter (weights) in filters last format - RSCK format where:
+    # R = height of the filter ,S = width of the filter, C = number of channels per filter,K = number of filters
+    # The default PyTorch convolution weight ordering is ‘filters first’ (KCRS).
+    # Therefore a re-ordering/permutation of all the convolution weights from KCRS to RSCK format is required
+    # before convolution operations
+    def permute_params(model: "pl.LightningModule", to_filters_last: bool) -> None:
+        """permute the params from filters first (KCRS) to filters last(RSCK) or vice versa.
+
+        and permute from RSCK to KCRS is used for checkpoint saving
+        """
+        with torch.no_grad():
+            for name, param in model.named_parameters():
+                if param.ndim == 4:
```

Comment:
what about other types of convolutions like 1D or 3D? This won't be catched at all and IMO this is relevant as well.
We cannot assume 2D operations only.

Also we should do some filtering to only apply this to weights of `_ConvNd`. If I have a custom model that for any reason has a 4D parameter that is never used for convolution, I would be pretty confused to have that one permuted like that. This is really error-prone and has to be regarded.

Reply:
Agreed @justusschock . We will be removing this need in the subsequent releases. The idea here it to showcase one example to users who might want to work on existing versions of synapse AI